drivers: video: video_stm32_dcmi: Use video buffers for DCMI buffer.

iabdalkader opened the pull request:
Instead of reserving a static (possibly unaligned) buffer for DCMI, this patch reserves and holds one of the video buffers to use as the main DCMI buffer. This buffer will be aligned (using the alignment specified in the config) and will either be allocated from `video_common` pool or a shared multi-heap (if enabled).

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -22,7 +22,17 @@
 
 LOG_MODULE_REGISTER(video_stm32_dcmi, CONFIG_VIDEO_LOG_LEVEL);
 
+#if defined(CONFIG_VIDEO_BUFFER_USE_SHARED_MULTI_HEAP)
+#include <zephyr/multi_heap/shared_multi_heap.h>
+#define VIDEO_STM32_HEAP_ALLOC(align, size, timeout)                                              \
+	shared_multi_heap_aligned_alloc(CONFIG_VIDEO_BUFFER_SMH_ATTRIBUTE, align, size)
+#define VIDEO_STM32_HEAP_FREE(block) shared_multi_heap_free(block)
+#else
 K_HEAP_DEFINE(video_stm32_buffer_pool, CONFIG_VIDEO_BUFFER_POOL_SZ_MAX);
+#define VIDEO_STM32_HEAP_ALLOC(align, size, timeout)                                              \
+	k_heap_aligned_alloc(&video_stm32_buffer_pool, align, size, timeout);
+#define VIDEO_STM32_HEAP_FREE(block) k_heap_free(&video_stm32_buffer_pool, block)
+#endif
```
It looks like this snippet is the same as here save for the symbol name:
https://github.com/zephyrproject-rtos/zephyr/blob/e4389a29214135a9f3004a3b8547d39e0d35cc21/drivers/video/video_common.c#L16-L24

How about moving the `video_common.c` macros (except `K_HEAP_DEFINE()`) to `video_common.h`, and adding a `buffer_pool` argument to specify which buffer pool is targeted?

That way, both `video_common.c` and `video_some_driver.c` can use the same implementation.

```suggestion
#if defined(CONFIG_VIDEO_BUFFER_USE_SHARED_MULTI_HEAP)
#include <zephyr/multi_heap/shared_multi_heap.h>
#define VIDEO_STM32_HEAP_ALLOC(buffer_pool, align, size, timeout)                                              \
	shared_multi_heap_aligned_alloc(CONFIG_VIDEO_BUFFER_SMH_ATTRIBUTE, align, size)
#define VIDEO_STM32_HEAP_FREE(buffer_pool, block) shared_multi_heap_free(block)
#else
K_HEAP_DEFINE(video_stm32_buffer_pool, CONFIG_VIDEO_BUFFER_POOL_SZ_MAX);
#define VIDEO_STM32_HEAP_ALLOC(buffer_pool, align, size, timeout)                                              \
	k_heap_aligned_alloc(buffer_pool, align, size, timeout);
#define VIDEO_STM32_HEAP_FREE(buffer_pool, block) k_heap_free(buffer_pool, block)
#endif
```

I am asking for your feedback on this, and if this does not make sense, the proposed solution also makes sense to me!

iabdalkader replied:
> How about moving the video_common.c macros (except K_HEAP_DEFINE()) to video_common.h,

Either way is fine with me. There's no `video_common.h` though, new file?

iabdalkader replied:
Ok. How about now?

josuah replied:
Thank you for taking the effort of refactoring this in generic APIs!

josuah reviewed (commented):
(context for other readers =>) It seems like the "continuous mode" of DCMI uses a different model where processing happens in the leap between two frames, rather than use a different buffer every time. That could be improved in the future maybe, but until then, allocating a buffer locally and `memcpy()`-ing it on every `vbuf->data` coming from the API seems like the way to go.

See inline comment for a proposal for how to integrate it... Thank you!

josuah reviewed (dismissed):
It is possible that other collaborators will have a better outlook of how to handle this.

At my level, it looks like the best way to implement it without completely reorganizing the video APIs (i.e. switch to `net_buf` or `rtio` for buffer management).

Thank you very much!

josuah reviewed (dismissed):
Good catch, I missed it.

ngphibang commented:
Just a first look, perhaps I missed some thing but I don't understand why you need to expose the `VIDEO_HEAP_ALLOC` macro (old name `VIDEO_COMMON_HEAP_ALLOC`) to use it i a driver ? Why don't just use the `video_buffer_alloc()` API ?

ngphibang commented on drivers/video/video_common.c:
```diff
@@ -10,17 +10,12 @@
-#define VIDEO_COMMON_HEAP_ALLOC(align, size, timeout)                                              \
-	k_heap_aligned_alloc(&video_buffer_pool, align, size, timeout);
-#define VIDEO_COMMON_FREE(block) k_heap_free(&video_buffer_pool, block)
+K_HEAP_DEFINE(VIDEO_BUFFER_POOL, CONFIG_VIDEO_BUFFER_POOL_SZ_MAX*CONFIG_VIDEO_BUFFER_POOL_NUM_MAX);
```
Should have some spaces

`K_HEAP_DEFINE(VIDEO_BUFFER_POOL, CONFIG_VIDEO_BUFFER_POOL_SZ_MAX * CONFIG_VIDEO_BUFFER_POOL_NUM_MAX);`

iabdalkader commented:
> Why don't just use the `video_buffer_alloc()` API ?

Because as it stands, `VIDEO_COMMON_HEAP_ALLOC` uses the pool defined in `video_common.c`, which is the main video buffers FIFO. `video_stm32` allocates another separate buffer, and copies it to this FIFO. If you have only one buffer in that FIFO `CONFIG_VIDEO_BUFFER_POOL_SZ_MAX=1`, and let `video_stm32` reserve it, it will never be able to dequeue another buffer to copy to.

ngphibang commented:
Ok, I understand the context now.

- 1. The DCMI driver needs an additional buffer for its personal usage (a part from the main video buffer pools used for the whole camera pipeline) => for this, I think some aspects need to be investiagted too, e.g does memcpy() used here is optimal, etc. But well, this is not in the scope of this PR.
- 2. This PR adds an option to allocate this additional buffer (required soly by the DCMI) either on the shared multi-heap (if `CONFIG_VIDEO_BUFFER_USE_SHARED_MULTI_HEAP` enabled) or on the dcmi driver heap.

So, I think you just need to do it in the DCMI driver. Refactoring the video_common to expose the macro VIDEO_HEAP_ALLOC is not the right way to do here because

- The VIDEO_HEAP_ALLOC and video_buffer_alloc() is to allocate buffers for the main pool, so it is on the main video_buffer_pool heap. The additional buffer needed by the DCMI is something related only to the DCMI not the main video buffer pool.
- Moreover, the refactoring does not help to save number of code lines.

Even if we want a "common" macro to allocate buffers on any heap, this should be defined system-wide, not inside the video subsystem (but again, it does not save much code lines)

iabdalkader commented:
> So, I think you just need to do it in the DCMI driver. Refactoring the video_common to expose the macro VIDEO_HEAP_ALLOC is not the right way to do here because

You mean duplicate the macro? The initial PR did that exactly, but then I was asked to move it to common header. Are we all in agreement this time that it needs to be duplicated?

Alternatively, requiring a minimum number of buffers (in the case of stm32 >= 2) for stm32 is another valid option, and then we can use `video_buffer_aligned_alloc`. The only drawback I see is that using `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX` from the application might be confusing as it will be `-1` for the DCMI.

ngphibang commented:
> You mean duplicate the macro? The initial PR did that exactly, but then I was asked to move it to common header. Are we all in agreement this time that it needs to be duplicated?

Sorry for that, cc @josuah (?). It is used only in one place so you **maybe** don't need to define the macro, just 

```
#if defined(CONFIG_VIDEO_BUFFER_USE_SHARED_MULTI_HEAP)
k_heap_alloc();
#else
shared_multi_heap_aligned_alloc()
```

but if it needs to be duplicated, we did it.

As explained above, `VIDEO_HEAP_ALLOC` or `video_buffer_alloc()` is to allocate buffer on the video buffer pool heap. If we want to generalize it for allocation of any particular heap, it should be an utility macro defined system wide.

There are examples such as a display driver needs a specific display heap for buffer allocation as [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_mcux_elcdif.c#L31) so it does not mean that display driver should use the refactored VIDEO_HEAP_ALLOC macro (?).

> Alternatively, requiring a minimum number of buffers (in the case of stm32 >= 2) for stm32 is another valid option, and then we can use video_buffer_aligned_alloc. The only drawback I see is that using CONFIG_VIDEO_BUFFER_POOL_NUM_MAX from the application might be confusing as it will be -1 for the DCMI.

In general, it could be an option. The DCMI can use the common video buffer pool for its personal purpose as well. The `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX` default value is already 2. But the problem is this config can be changed accidentally / easily by the application and the application does not know that the DCMI driver needs an additional buffer.

A similar option which was discussed [here](https://github.com/zephyrproject-rtos/zephyr/pull/69250) is to use the system heap instead of the driver heap and extend it in the driver with CONFIG_HEAP_MEM_POOL_ADD_SIZE_XXX but if I remember well it does not add the additiobnal required size but the specify the whole size (the discussion was a long time ago, I need to recall it).

iabdalkader commented:
> so it does not mean that display driver should use the refactored VIDEO_HEAP_ALLOC macro (?).

Actually it should, because it reserves a chunk of memory that it may not use at all.

> But the problem is this config can be changed accidentally / easily by the application and the application does not know that the DCMI driver needs an additional buffer.

I thought about this and have a better idea. What if `video_stm32_dcmi.c` dequeued a video buffer instead of allocating raw memory? The application allocates video buffers with `video_buffer_aligned_alloc` and enqueues them with `video_enqueue`. Then when `video_stm32_dcmi_stream_start` is called it will dequeue one video buffer and hold it until streaming is stopped. Note that the driver will fail to build if the MAX buffers is less than 2 for stm32 video.

iabdalkader committed:
drivers: video: video_stm32_dcmi: Use video buffers for DCMI buffer.

Instead of reserving a static (possibly unaligned) buffer for DCMI,
this patch reserves and holds one of the video buffers to use as the
main DCMI buffer. This buffer will be aligned (using the alignment
specified in the config) and will either be allocated from `video_common`
pool or a shared multi-heap (if enabled).

Signed-off-by: Ibrahim Abdalkader <i.abdalkader@gmail.com>

ngphibang commented:
> so it does not mean that display driver should use the refactored VIDEO_HEAP_ALLOC macro (?).
> 
> Actually it should, because it reserves a chunk of memory that it may not use at all.

I think you misunderstood my point here. I mean the display driver could use a common macro like HEAP_ALLOC() defined system wide (to save some line of codes) but it should not include **video_common.h** and use VIDEO_HEAP_ALLOC() because VIDEO_HEAP_ALLOC() is for video subsystem.

To be clearer, take a look at the signature of newly refactored VIDEO_HEAP_ALLOC:

`#define VIDEO_HEAP_ALLOC(buffer_pool, align, size, timeout) `

there is nothing specific to **_video_** anymore, so that's why it should be HEAP_ALLOC(buffer_pool, align, size, timeout) defined system wide so that any video / display / others drivers could use it.

ngphibang commented:
Thanks. I need some time to understand the context / purpose of the additional buffer in the DCMI driver (I missed the review of this driver). Otherwise, @CharlesDias may have a better look.

iabdalkader commented:
> I need some time to understand the context / purpose of the additional buffer in the DCMI driver

No worries. FWIW, there's no reason not to use the FIFO to pass buffers between driver application without `memcpy` (i.e., put back the buffer after capture, and get another one). However, if the user/application does not dequeue, process and enqueue fast enough the FIFO the DCMI side will underrun and will have to stop streaming, and possibly restart later. This, I believe, is the only reason for the additional buffer. Is it fixable? Probably yes, but again just trying to fix my immediate problem :)

josuah commented:
Thanks all for this in-depth study!

> One way to fix this is to require a minimum of two buffers for this driver, but I don't know how to do that, or if it's even possible, is it?

It is possible to declare it, although, not taken advantage of through the samples or API:

https://github.com/zephyrproject-rtos/zephyr/blob/9e08560f0824769b658e418c1a18702330ef9e85/include/zephyr/drivers/video.h#L98

> maybe don't need to define the macro, just `#if [...] shared_multi_heap_aligned_alloc(); #else k_heap_alloc();`

That sounds good, it would be a more lightweight intermediate solution for a more general fix coming later.

> What if video_stm32_dcmi.c dequeued a video buffer instead of allocating raw memory? The application allocates video buffers with video_buffer_aligned_alloc and enqueues them with video_enqueue.

That is how most video drivers work it seems, and if it is possible to make DCMI work that way, that avoids the problem of local video memory allocation in the driver.

> context / purpose of the additional buffer in the DCMI driver

The way the buffers are currently loaded/unloaded:

https://github.com/zephyrproject-rtos/zephyr/blob/35abb312849efbd496542d1c6805aa2b35ca0db6/drivers/video/video_stm32_dcmi.c#L68-L83

I do not see any API call that allows to update that `pData` to the buffer used for I/O:

https://github.com/zephyrproject-rtos/hal_stm32/blob/1d1f81866ccbaa6e84e9960ed763e005d1e45560/stm32cube/stm32f7xx/drivers/include/stm32f7xx_hal_dcmi.h#L559

`HAL_StatusTypeDef HAL_DCMI_Start_DMA(DCMI_HandleTypeDef *hdcmi, uint32_t DCMI_Mode, uint32_t `**`pData`**`, uint32_t Length);`

Maybe the hardware does not allow to update the buffer without stopping/starting the engine?

iabdalkader commented:
> That is how most video drivers work it seems, and if it is possible to make DCMI work that way, that avoids the problem of local video memory allocation in the driver.

I've already implemented this and updated the PR, no local video buffer allocation but still uses `memcpy` as before, this has not changed.

> Maybe the hardware does not allow to update the buffer without stopping/starting the engine?

The DMA does allow updating the target address when not in use, general or just in double buffer mode can't remember,  but either way this is a different issue, which I'm not trying to fix.

josuah reviewed (approved):
I missed the update. Very good first step indeed!
I prefer it over what I originally suggested (previous LGTM) as I did not think of this idea.
+1 for my part!

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -244,16 +246,16 @@ static int video_stm32_dcmi_stream_start(const struct device *dev)
-	data->buffer = k_heap_alloc(&video_stm32_buffer_pool, buffer_size, K_NO_WAIT);
-	if (data->buffer == NULL) {
-		LOG_ERR("Failed to allocate DCMI buffer for image. Size %d bytes", buffer_size);
+	data->vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT);
+
+	if (data->vbuf == NULL) {
+		LOG_ERR("Failed to dequeue a DCMI buffer.");
```
After this commit, if a video buffer is not passed before start(), this would return an error.
I think that all examples in `main` would still work:

here: 2 buffers
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/video/tcpserversink/src/main.c#L41
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/video/capture_to_lvgl/src/main.c#L22

or otherwise, `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX`
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/video/capture/src/main.c#L85

iabdalkader replied:
Yes, the applications should still work, mine still works without changes:

App: allocates 2 buffers.
App: enqueues 2 buffers in fifo_in.
App: starts streaming.
Driver: dequeues buffer 1 from fifo_in and holds it for use as DCMI buffer.
Driver: Frame ready, dequeue buffer 2 from fifo_in, memcpy, enqueue in fifo_out.

CharlesDias commented:
Hi, @iabdalkader. Thank you for your contribution! This is a great improvement! :)

@ngphibang and @josuah, is there another way to release the video buffers besides using the `video_buffer_release`? I'm wondering if there's a possibility for the video buffer to be released while the DCMI driver is running. I believe this isn't possible under normal usage conditions.

Additionally, I tested the c[apture_to_lvgl ](https://docs.zephyrproject.org/latest/samples/drivers/video/capture_to_lvgl/README.html#video-capture-to-lvgl) sample on [MiniSTM32H743](https://docs.zephyrproject.org/latest/boards/weact/mini_stm32h743/doc/index.html#mini_stm32h743) and still worked. 
It has my +1!

josuah commented:
Thank you for testing it!

> is there another way to release the video buffers besides using the video_buffer_release?

If the application decides to use a `vbuf` after it was `enqueued()`, and before it was `dequeued()`, AFAIU, it is an undefined behavior (as the buffer is expected to be processed by the driver). It is expected to forget a `vbuf` pointer as soon as it is passed to `enqueue()`.

AFAICT, the current APIs give the driver the freedom of how many buffers it holds and in which order they are released.

https://github.com/zephyrproject-rtos/zephyr/blob/e4389a29214135a9f3004a3b8547d39e0d35cc21/drivers/video/video_common.c#L71

The video buffer variables are not exported, there is no way to access it from the outside other than the provided functions.

ngphibang reviewed (approved):
@CharlesDias @iabdalkader  I believe the `memcpy()` is not necessary and it does reduce the performance of the camera pipeline. To avoid this software copy, the DCMI driver (zephyr driver and / or the HAL driver) could implement a double buffering mechanism (similar to the NXP CSI drivers) so that when the DCMI dma-ing the image data to one buffer, the application can use another buffer for displaying.

But well, this is the problem of the original code. The current PR does not affect this issue and LGTM.

loicpoulain commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -276,8 +278,8 @@ static int video_stm32_dcmi_stream_stop(const struct device *dev)
-	/* Release the buffer allocated in stream_start */
-	k_heap_free(&video_stm32_buffer_pool, data->buffer);
+	/* Release the video buffer allocated in stream_start */
+	k_fifo_put(&data->fifo_in, data->vbuf);
```
Shouldn't we 'release' it after HAL_DCMI_stop? to be sure no frame event can happen in between?

iabdalkader replied:
This just puts the video buffer back on the queue, so even if a frame event occurs before the stop, the pointer should still be valid, so it's actually slightly better now because before it was free'd memory. I don’t mind making this fix, but if it will require approvals again, I’d prefer to send it in another PR. Note that this is where the old buffer was released.

loicpoulain replied:
Ok, would be good to have a following patch, This is equivalent to disowning the buffer, you should not use anymore after that and can not speculate on what is going to happen to it.

iabdalkader replied:
https://github.com/zephyrproject-rtos/zephyr/pull/84670
